### ISSUE TITLE
Send Buffer/binary response to client

### ISSUE THREAD

**Issue by sundowndev** [lukeautry/tsoa#800]
https://github.com/lukeautry/tsoa/issues/435 and #44  were both closed but the issue is still here. Is there any way to send files to client using tsoa? Why doesn't `@Response` decorator exists ?

## Sorting

- **I'm submitting a ...**

  - [x] bug report
  - [ ] feature request
  - [ ] support request

- I confirm that I
  - [x] used the [search](https://github.com/lukeautry/tsoa/search?type=Issues) to make sure that a similar issue hasn't already been submit

## Expected Behavior

Router to send file buffer to be downloaded by the client.

## Current Behavior

When using Attempt 1: the file is sent but the file is invalid and cannot be opened.

When using Attempt 2:

```
TypeError: Cannot read property 'on' of undefined
    at ReadStream.Readable.pipe (_stream_readable.js:669:8)
    at OrderETicketsController.read (/app/dist/controllers/orderEtickets/controllers.js:22:24)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
error Command failed with signal "SIGINT".
```

#### Attempt 1

```ts
this.setStatus(200);
this.setHeader('Content-Type', 'application/pdf');
this.setHeader('Content-Disposition', `attachment; filename=${fileName}`);

return fs.promises.readFile(filePath);
```

#### Attempt 2

```ts
req.res.download(filePath) // req.res is undefined!
```

## Context (Environment)

Version of the library: 3.3.0 (latest)
Version of NodeJS: v12.18.1

- Confirm you were using yarn not npm: [x]

**Comment by sundowndev** [comment]
Ok I finally managed to do it.

The generated routes file already handles streams, you just have to return a valid readable stream.

```typescript
function returnHandler(response: any, statusCode?: number, data?: any, headers: any = {}) {
    Object.keys(headers).forEach((name: string) => {
      response.set(name, headers[name]);
    });

    if (data && typeof data.pipe === 'function' && data.readable && typeof data._read === 'function') {
      // Here, data.pipe must be a function
      data.pipe(response);

    } else if (data || data === false) { // === false allows boolean result
      response.status(statusCode || 200).json(data);
    } else {
      response.status(statusCode || 204).end();
    }
  }
```

Controller: 

```typescript
public async download(): Promise<fs.ReadStream> {
      const filePath: string = '/path/to/file';
      const fileName: string = 'doc.pdf';

      const stat: fs.Stats = await fs.promises.stat(filePath);

      this.setStatus(200);
      this.setHeader('Content-Type', mime.lookup(filePath));
      this.setHeader('Content-Length', stat.size.toString());
      // Removing this line will cause to not launch the download, just serve the file as it
      this.setHeader('Content-Disposition', `attachment; filename=${fileName}`);

      return fs.createReadStream(filePath);
  }
```

The solution is more complicated than using `res.download` but it works. I'd suggest to implement a feature to make file download easier.

**Comment by WoH** [comment]
I assume this affects the Express template only?
We are checking for Readable Streams, seems like we should handle Buffer aswell, the type resolution seems fine, but can't tell from SwUi, please post the relevant spec.

**Comment by sundowndev** [comment]
Here's an example of output from tsoa :) 

```json
		"/download": {
			"get": {
				"operationId": "downloadFile",
				"responses": {
					"200": {
						"description": "OK",
						"content": {
							"application/json": {
								"schema": {
									"type": "string",
									"format": "byte"
								}
							}
						}
					},
				},
				"description": "Download a file.",
				"tags": [
					"Download"
				],
				"security": [],
				"parameters": []
			}
		},
```

**Comment by mb21** [comment]
Adding something like this to the default express template should fix it:

```js
import { Readable } from 'stream';

...

} else if (data instanceof Buffer) {
  Readable.from(data).pipe(response);
```

**Comment by Sawtaytoes** [comment]
I dunno if this is related, but I saw the code for `.readable` in `routes.ts` and don't think that's quite enough.

I'm using S3's `GetObjectCommand`, and it returns me the "body" of an object like so:
```ts
return (
  object
  .Body
  .transformToByteArray()
)
```

When I do `transformToByteArray`, I can send that as a JSON-stringified Uint8Array to the client, but that's causing issues for me.

Instead, I had to modify the response this way:

```ts
new FileService()
.get({
  accountId,
  filename,
})
.then((
  fileData
) => {
  request
  ?.res
  ?.write(
    fileData
  )

  request
  ?.res
  ?.end()
})
```

Is it possible there's something I'm missing that's not documented? There's a whole section on uploading files to tsoa (which works for me), but I'm missing the part about how to download files from tsoa.
